Fix regression with 'dynamic' or 'NSManaged' properties witnessing protocol requirements

slavapestov opened the pull request:
The 'modify' coroutine should always call getter and setter methods, and not directly access storage. It should also be marked serialized so that it can be referenced from serialized protocol witness thunks.

Fixes <https://bugs.swift.org/browse/SR-8657> / <rdar://problem/43951732>.

slavapestov committed:
Sema: Simplify accessor synthesis a bit

slavapestov committed:
AST: Fix getAccessSemanticsFromContext() for 'modify' coroutines of 'dynamic' properties

Normally we want accessors to perform direct-to-storage access on
their own storage, otherwise we can't ever implement the accessor;
but one special case is the 'modify' coroutine of a 'dynamic'
property, which has to dynamically dispatch to the getter and
setter methods, since they may be replaced at runtime.

First part of <https://bugs.swift.org/browse/SR-8657> /
<rdar://problem/43951732>.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -1392,22 +1392,43 @@ static bool isPolymorphic(const AbstractStorageDecl *storage) {
+  // must use ordinary semantics, so that the 'modify' coroutine for a
+  // 'dynamic' property uses Objective-C message sends and not direct
+  // access to storage.
+  if (accessor->getAccessorKind() == AccessorKind::Modify &&
```
Rather than checking for `modify`, this should probably include every accessor other than `get` or `set`.

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -1392,22 +1392,43 @@ static bool isPolymorphic(const AbstractStorageDecl *storage) {
+  // must use ordinary semantics, so that the 'modify' coroutine for a
+  // 'dynamic' property uses Objective-C message sends and not direct
+  // access to storage.
+  if (accessor->getAccessorKind() == AccessorKind::Modify &&
```
Rather than checking for `modify`, this should probably include every accessor other than `get` or `set`. (It'd be a problem if we synthesized `read` as well.)

slavapestov replied:
Yeah. It would be enough to just check hasForcedStaticDispatch() - which also needs to be renamed. It means “I don’t have a vtable entry, I have shared linkage, and I dynamically dispatch getters and setters”. It’s only set for a read or modify on a dynamic property (for now at least). Got any suggestions for a better name?

jrose-apple reviewed (commented):
Looking pretty good to me!

slavapestov commented:
Oops, CoreStoreDemo still fails in the same way. Looks like there’s another bug here. Also I need to update a couple of SILGen tests. Updated version coming up, but I don’t expect it to change much.

slavapestov commented:
It passes locally. I wonder if PR testing built the wrong revision?

slavapestov commented:
.... ah, no, it's -enable-testing.

slavapestov committed:
SIL: Accessors synthesized on-demand are serialized when visible outside the module

Second part of <https://bugs.swift.org/browse/SR-8657> /
<rdar://problem/43951732>.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -1392,22 +1392,42 @@ static bool isPolymorphic(const AbstractStorageDecl *storage) {
+  // dispatch must use ordinary semantics, so that the 'modify' coroutine for a
+  // 'dynamic' property uses Objective-C message sends and not direct access to
+  // storage.
+  if (accessor->hasForcedStaticDispatch())
```
We should AST-verify that getters and setters never have this set if we don't already.

slavapestov replied:
I'll do that in a follow-up PR.

jrose-apple replied:
Yeah, sorry, didn't mean to tack it on to this one.

slavapestov commented:
Ok, CoreStore is now passing!

```
17:06:56 Failures:
17:06:56   FAIL: Chatto, 4.0.3, 3e4b1a, ChattoApp, generic/platform=iOS
17:06:56   FAIL: Chatto, 4.0.3, 3e4b1a, ChattoAdditions, generic/platform=iOS
```

@DougGregor remaining failure is on you :)

rjmccall reviewed (commented):
LGTM.
